In [27]:
import pandas as pd 
import numpy as np 
from sklearn.metrics import mean_absolute_error, r2_score, mean_absolute_percentage_error

In [28]:
dataset = pd.read_csv('student_scores.csv')

In [29]:
X = dataset.iloc[:, :-1].values 
y = dataset.iloc[:, 1].values

In [30]:
from sklearn.model_selection import train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [31]:
from sklearn.linear_model import LinearRegression 
regressor = LinearRegression() 
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)

In [32]:
print(regressor.intercept_)
print(regressor.coef_)

2.018160041434662
[9.91065648]


In [33]:
print(f"MAE: {mean_absolute_error(y_test, y_pred)}")
print(f"R2: {r2_score(y_test, y_pred)}")
print(f"MAPE: {mean_absolute_percentage_error(y_test, y_pred)}")

MAE: 4.183859899002982
R2: 0.9454906892105354
MAPE: 0.12568891617045685


Модель обладает хорошей точностью. R2 -> 1, MAPE < 20%, MAE ~ 4 хорошее значение для нашей модели.

In [34]:
from sklearn import datasets
diabetes = datasets.load_diabetes()
df = pd.DataFrame(diabetes.data, columns = diabetes.feature_names)
df['target'] = diabetes.target

In [35]:
X = df["bmi"].values.reshape(-1,1)
y = df["target"].values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [36]:
regressor = LinearRegression() 
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)

In [37]:
print(regressor.intercept_)
print(regressor.coef_)

152.28824927379569
[981.65543614]


In [38]:
print(f"MAE: {mean_absolute_error(y_test, y_pred)}")
print(f"R2: {r2_score(y_test, y_pred)}")
print(f"MAPE: {mean_absolute_percentage_error(y_test, y_pred)}")

MAE: 52.94370285288119
R2: 0.19057346847560108
MAPE: 0.4516904624956549


Модель имеет плохую точность. R2 не близко к 1, MAE ~ 53 - довольно заметная ошибка, но приемлимая. MAPE ~ 45% указывает на то, что модель ошибается почти на половину значения target.
На графике видно, что четкой плотной корреляции между BMI и target нет.